g3doc/Tutorials/Toy/Ch-2.md

```diff
@@ -23,139 +23,320 @@ addresses this issue by being designed for extensibility. As such, there are
+operations in MLIR is extensible. This means that the infrastructure must be
+able to opaquely reason about the structure of an operation. This is done by
+boiling down the composition of an operation into discrete pieces:
+
+-   A name for the operation.
+-   A source location for debugging purposes.
+-   A list of SSA operand values.
+-   A list of [types](../../LangRef.md#type-system) for result values.
+-   A list of [attributes](../../LangRef.md#attributes).
+-   A list of successors [blocks](../../LangRef.md#blocks) (for branches
+    mostly).
+-   A list of [regions](../../LangRef.md#regions) (for structural operations
+    like functions).
 
 Finally, in MLIR every operation has a mandatory source location associated with
 it. Contrary to LLVM where debug info locations are metadata and can be dropped,
 in MLIR the location is a core requirement which translates in APIs manipulating
 operations requiring it. Dropping a location becomes an explicit choice and
 cannot happen by mistake.
 
-## Opaque API
+### Opaque API
 
-MLIR is designed to be a completely extensible system, as such the
-infrastructure has the capability to opaquely represent operations (as well as
-attributes, types, etc.) that have not been registered. This allows MLIR to
-parse, represent, and round-trip any valid IR. For example, the following can
-round-trip through *mlir-opt*:
+MLIR is designed to be a completely extensible system, and as such, the
+infrastructure has the capability to opaquely represent all of its core
+components: attributes, operations, types, etc. This allows MLIR to parse,
+represent, and round-trip any valid IR. For example, we could inject our toy
+operation from above into an .mlir file and round-trip through *mlir-opt*
+without registering anything:
 
-```MLIR(.mlir)
-func @some_func(%arg0: !random_dialect<"custom_type">) -> !another_dialect<"other_type"> {
-  %result = "custom.operation"(%arg0) { attr = #random_dialect<"custom_attribute"> } : (!random_dialect<"custom_type">) -> !another_dialect<"other_type">
-  return %result : !another_dialect<"other_type">
+```mlir
+func @toy_func(%tensor: tensor<2x3xf64>) -> tensor<3x2xf64> {
+  %t_tensor = "toy.transpose"(%tensor) { inplace = true } : (tensor<2x3xf64>) -> tensor<3x2xf64>
+  return %t_tensor : tensor<3x2xf64>
 }
 ```
 
-Here MLIR will enforce some structural constraints (SSA, block termination,
-etc.) but otherwise the types and the `custom.operation` are completely opaque.
+In the cases of unregistered attributes, operations, types, MLIR will enforce
+some structural constraints (SSA, block termination, etc.) but otherwise they
+are completely opaque. This can be useful for bootstrapping purposes, but it is
+generally advised against. Opaque operations must be treated conservatively by
+transformations and analyses, cannot be properly verified for custom semantics,
+and are much harder to construct and manipulate.
```

Comment:
I find this quite underwhelming since this is intended to replace a very large piece of what Ch2 was intended to. It loses the motivation for dialect and registration by papering over the problem. I find this to make the chapter less accessible to beginners (which was the point of having two separate chapters). Can you re-read around line 201 in the original Ch2 to get the spirit?